tests/unit/test_gcp_clients.py

```diff
@@ -0,0 +1,250 @@
+                                {'key': 'default', 'value': 'true'}
+                            ],
+                        },
+                    }],
+                    'warning': {
+                        'warning': 'object'
+                    }
+                }
+            },
+        }
+
+    @pytest.mark.parametrize('query_str,instance_meta,log_call_count', [
+        # test one page of results + zone filter
+        ('maxResults=10&filter=zone eq us-west9-z', None, 0),
+        # test instance filtering when filtered by tags
+        ('maxResults=10&filter=', ['bl-tag'], 1),
+        # test instance filtering when filtered by metadata
+        ('maxResults=10&filter=', {'key': 'type', 'value': 'c0ffee-b33f'}, 1),
+    ])
+    @pytest.mark.asyncio
+    async def test_list_instances(
+            self, compute_rsp, patch_compute_base_url, get_gce_client, caplog,
+            query_str, instance_meta, log_call_count):
+        """Client uses multiple filters to process results."""
+        caplog.set_level(logging.INFO)
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+
+        if instance_meta:
+            blacklisted_instance = \
+                copy.deepcopy(
+                    compute_rsp['items']['us-west1-z']['instances'][0])
+            blacklisted_instance['name'] = 'instance-2'
+            compute_rsp['items']['us-west1-z']['instances'].append(
+                blacklisted_instance)
+
+        if isinstance(instance_meta, list):
+            blacklisted_instance['tags']['items'] = instance_meta
+            gce_client.blacklisted_tags.append(instance_meta[0])
+        elif isinstance(instance_meta, dict):
+            blacklisted_instance['metadata']['items'].append(instance_meta)
+            blacklisted_metadata = {
+                instance_meta['key']: instance_meta['value']
+            }
+            gce_client.blacklisted_metadata.append(blacklisted_metadata)
+
+        with aioresponses() as m:
+            filter_url = (f'{patch_compute_base_url}{query_str}')
+            m.get(filter_url, payload=compute_rsp)
+
+            kwargs = {}
+            if instance_meta is None:
+                kwargs = {'instance_filter': query_str[query_str.find('zone'):]}
+
+            results = await gce_client.list_instances(
+                'test-project',
+                page_size=10,
+                **kwargs)
+
+        expected_results = [{
+            'hostname': 'instance-1',
+            'internal_ip': '192.168.0.1',
+            'external_ip': '1.1.1.1'
+        }]
+        assert results == expected_results
+        assert len(caplog.records) == log_call_count
+
+    @pytest.mark.asyncio
+    async def test_list_instances_bad_json(
+            self, compute_rsp, patch_compute_base_url, get_gce_client, caplog):
+        """Client ignores incomplete API replies."""
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+        caplog.set_level(logging.INFO)
```

Comment:
Maybe we should have a set of logging fixtures, like `caplog_info`, `caplog_debug`, `caplog_warn` so we don't override `caplog`, but also don't have to set it all the time 🤔

Reply:
I'd prefer to leave this for a later PR.